### ISSUE TITLE
Python UDFs marked as side_effects = False are not having results cached/deduplicated when called from SELECT

### ISSUE THREAD

**Issue by williamkaper** [duckdb/duckdb#13738]
### What happens?

When you call a UDF from a query's select statement and mark it as not having side_effects, the function only appears to be getting cached/deduplicated when the parameters are literals. When you try to pass in values from columns in the select statement, even if for n rows that value is the exact same value, the function is called and evaluated n times. This kills performance when you use it in a query/model that calls the function to transform a value for a result set and the result set has a finite of combinations. 

The where clause, on the other hand, does cache the result even when there are duplicate values in the table being queried. 

If would be great if DuckDB optimized the calls on the select side as well using whatever caching/deduplication mechanism is being used for calling UDFs in the where clause. 

### SAMPLE UDF

```python
import duckdb
from duckdb import DuckDBPyConnection
from dbt.adapters.duckdb.plugins import BasePlugin
from dbt.adapters.duckdb.utils import TargetConfig

def immutable_test(arg1: str) -> bool:
    print(arg1)
    return True

class Plugin(BasePlugin):
    def configure_connection(self, conn: DuckDBPyConnection):
        conn.create_function("immutable_test", immutable_test, [str], bool, side_effects = False)

    def store(self, target_config: TargetConfig):
        assert target_config.config.get("key") == "value"
```

### SAMPLE MODEL IN DBT where you can see the duplicate calls (UDF CALLED FROM SELECT)

```sql
SELECT
  immutable_test(a.name) AS udf_result
FROM duckdb_settings() as A
CROSS JOIN duckdb_settings() as B
```

### SAMPLE MODEL IN DBT where it only calls a unique number of times (UDF CALLED FROM THE WHERE CLAUSE)

```sql
SELECT
  immutable_test(a.name) AS udf_result
FROM duckdb_settings() as A
CROSS JOIN duckdb_settings() as B
```

### OS:

Ubuntu Linux x64

### DuckDB Version:

1.0.0

### DuckDB Client:

Python through DBT

### Full Name:

William Kaper

### Affiliation:

HubiFi Corporation

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a stable release

### Did you include all relevant data sets for reproducing the issue?

Yes

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by Tishj** [comment]
I wonder if this can't be fixed by the memoization decorator (I believe that's in a builtin python module (?)), as we're just calling the method like you would call it normally through python

The memoization decorator looks at the input data and should also function here unless I'm mistaken

**Comment by williamkaper** [comment]
> I wonder if this can't be fixed by applying the the memoization decorator to your python method (I believe that's in a builtin python module (?)), as we're just calling the method like you would call it normally through python
> 
> The memoization decorator looks at the input data and should also function here unless I'm mistaken (https://docs.python.org/3/library/functools.html)

I don't know how python's interpreter works or how DuckDB calls out to it, but would that alleviate the cost of the call out to python, or just the computation cost for the UDFs internal operations? I'm assuming DuckDB would still eat the interop call out to python, but I'd love for a DuckDB eng to weigh in

**Comment by Tishj** [comment]
That would only alleviate the computation cost for the UDFs internal operations

**Comment by williamkaper** [comment]
It's still a smart optimization to add in any case. It would be great if the DuckDB guys added a wrapper for it to a method decorator that allows you to specify through the decorator if the method is immutable (in PostgreSQL parlance) which would be an alias for @cache.

**Comment by hannes** [comment]
I agree it is a smart optimization, depending on the execution cost of the function. Implementing this is currently not on the roadmap but we might get to it in the future. If this is a priority, we are happy to discuss a commercial arrangement to prioritize implementation. Feel free to reach out at hannes@duckdblabs.com

**Comment by williamkaper** [comment]
We are working with the motherduck guys to see if we can migrate our workloads from PostgreSQL to DuckDB/MotherDuck. I think we can work around this issue for now on our side by moving one of our UDFs to a DBT Macro and changing the other to dedup the calls through some workarounds with filtered CTEs, but it would be nice if this were addressed for a future enhancement in the platform so UDFs can reach their full potential.
